cranelift/codegen/src/opts/algebraic.isle

```diff
@@ -204,6 +200,17 @@
                       (iconst _ k)))
       (band ty x (iconst ty (imm64_ushr ty (imm64 (ty_mask ty)) k))))
 
+;; For signed shifts, `(x << k) >> k` does sign-extension from `n` bits to
+;; `n+k` bits. In the special case where `x` is the result of either
+;; `sextend` or `uextend` from `n` bits to `n+k` bits, we can implement this
+;; using `sextend`.
+(rule (simplify (sshr wide (ishl _ (uextend _ x @ (value_type narrow)) (iconst _ shift)) (iconst _ shift)))
+      (if-let shift (imm64 (u64_sub (ty_bits_u64 wide) (ty_bits_u64 narrow))))
+      (sextend wide x))
+(rule (simplify (sshr wide (ishl _ x @ (sextend _ (value_type narrow)) (iconst _ shift)) (iconst _ shift)))
```

Comment:
Just for clarity, even though it's strictly redundant given CLIF typechecking, could we write `wide` for the wildcards for the type params to `ishl` and `sextend` here? (General comment across the rules, and we may not have been consistent elsewhere with this; but especially for rules operating across multiple types, it makes it much easier to reason about!)